sw/device/tests/rom_ext/rom_ext_parser_unittest.cc

```diff
@@ -0,0 +1,185 @@
+
+TEST_F(ImageLengthGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_IMAGE_LENGTH_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_image_len_get(base_addr_), 0xa5a5a5a5);
+}
+
+class ImageVersionGetTest : public ParserTest {};
+
+TEST_F(ImageVersionGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_IMAGE_VERSION_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_version_get(base_addr_), 0xa5a5a5a5);
+}
+
+class ImageTimestampGetTest : public ParserTest {};
+
+TEST_F(ImageTimestampGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_IMAGE_TIMESTAMP_OFFSET, 0xcdcdcdcd);
+  EXPECT_READ32(ROM_EXT_IMAGE_TIMESTAMP_OFFSET + sizeof(uint32_t), 0xabababab);
+
+  uint64_t result = rom_ext_timestamp_get(base_addr_);
+  EXPECT_EQ(result, 0xababababcdcdcdcd);
+}
+
+class AlgorithmIdGetTest : public ParserTest {};
+
+TEST_F(AlgorithmIdGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_SIGNATURE_ALGORITHM_IDENTIFIER_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_algorithm_id_get(base_addr_), 0xa5a5a5a5);
+}
+
+class ExponentGetTest : public ParserTest {};
+
+TEST_F(ExponentGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_EXPONENT_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_exponent_get(base_addr_), 0xa5a5a5a5);
+}
+
+class UsageConstraintsGetTest : public ParserTest {};
+
+TEST_F(UsageConstraintsGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_USAGE_CONSTRAINTS_OFFSET, 0xa5a5a5a5);
+  EXPECT_EQ(rom_ext_usage_constraints_get(base_addr_), 0xa5a5a5a5);
+}
+
+class PeripheralLockdownInfoGetTest : public ParserTest {
+ protected:
+  PeripheralLockdownInfoGetTest() {
+    for (uint32_t i = 0; i < ROM_EXT_PERIPHERAL_LOCKDOWN_INFO_SIZE_WORDS; ++i) {
+      src_.data[i] = std::uniform_int_distribution<uint32_t>()(gen_);
+    }
+
+    base_addr_.base = &src_.data[0];
+  }
+
+  rom_ext_lockdown_info_t src_;
+
+ private:
+  std::random_device rd;
+  std::mt19937 gen_ = std::mt19937(rd());
+};
+
+TEST_F(PeripheralLockdownInfoGetTest, Success) {
+  rom_ext_lockdown_info_t dst;
+
+  EXPECT_EQ(rom_ext_peripheral_lockdown_info_get(base_addr_, &dst), true);
+  EXPECT_THAT(src_.data, ContainerEq(dst.data));
+}
+
+class PublicKeyGetTest : public ParserTest {
+ protected:
+  PublicKeyGetTest() {
+    for (uint32_t i = 0; i < ROM_EXT_SIGNATURE_PUBLIC_KEY_SIZE_WORDS; ++i) {
+      src_.data[i] = std::uniform_int_distribution<uint32_t>()(gen_);
+    }
+
+    base_addr_.base = &src_.data[0];
+  }
+
+  rom_ext_public_key_t src_;
+
+ private:
+  std::random_device rd;
+  std::mt19937 gen_ = std::mt19937(rd());
+};
+
+TEST_F(PublicKeyGetTest, Success) {
+  rom_ext_public_key_t dst;
+
+  EXPECT_EQ(rom_ext_public_key_get(base_addr_, &dst), true);
+  EXPECT_THAT(src_.data, ContainerEq(dst.data));
+}
+
+class ExtensionGetTest : public ParserTest {};
+
+TEST_F(ExtensionGetTest, Success) {
+  EXPECT_READ32(ROM_EXT_EXTENSION0_OFFSET_OFFSET, 0xaaaaaaaa);
+  EXPECT_READ32(ROM_EXT_EXTENSION0_CHECKSUM_OFFSET, 0xbbbbbbbb);
+
+  rom_ext_extension_t ext0 = rom_ext_extension0_get(base_addr_);
+  EXPECT_EQ(ext0.offset, 0xaaaaaaaa);
```

Comment:
Personally I would have defined operator== and operator<< for `rom_ext_extension_t`, since that would allow doing a single EXPECT, but that's just me. (My i2c DIF has examples of how to do this.)

Reply:
Went with Alphans suggestion.

Done.

Reply:
I missed this because it got merged before I could re-review, but naming matchers after `testing` matchers is a really bad idea. This should have been named something else. (Also, you probably should have used `MATCHER_P`...)

Reply:
Thanks for pointing these out mcy.
Naming: Sure, better to use a more specific name but I don't think there is a real problem with this.
`MATCHER_P`: AFAIK it won't give field-specific error information as `Field` does.